pilot/pkg/proxy/envoy/v2/eds.go

```diff
@@ -523,6 +523,7 @@ func endpointDiscoveryResponse(loadAssignments []*xdsapi.ClusterLoadAssignment,
 	}
 	for _, loadAssignment := range loadAssignments {
 		resource := util.MessageToAny(loadAssignment)
+		resource.TypeUrl = typeURL
```

Comment:
will this cause v3 responses being sent in response to v2 requests?

Reply:
No. We lock the version based on the first request - see `handleTypeURL`. So a given connection will always be v2 or v3 (per type) for its entire connection

Reply:
specifically `typeURL` here is derived from the request, so this is always making request == response

Reply:
Thank you.
Apologies for my unfamiliarity with Pilot, but i'm asking this because its relevant to the control plane we operate and I'm looking to use this as an example.
Will there be ever cases where due to rolling update half of envoy sidecars are on v2 and half on v3? 

imo v3 sidecars will also need v3 grpc handler to be implemented alongside the v2 [handler](https://github.com/istio/istio/blob/master/pilot/pkg/proxy/envoy/v2/ads.go#L174).

Reply:
> imo v3 sidecars will also need v3 grpc handler to be implemented alongside the v2 handler.

Yes that will be another PR. The transport protocol version is orthogonal to the resource versions. In the future we will add the other handler as well.

We will have cases where half are v2 and half are v3. however, a given proxy will be one or the other - it will never be mixed. So those v2 proxies will request v2 and get v2 responses, and the v3 proxies will request v3 and get v3 responses.

Reply:
Makes sense. When implementing the handler probably i will have some more questions about best practices for handling common code from 2 different handlers. Since we deal with explicit [v2 resources](https://github.com/istio/istio/blob/master/pilot/pkg/proxy/envoy/v2/ads.go#L23-L25) , i would love to learn how to manage with least code duplication.

Reply:
For the handler that part will probably need to be duplicated. Its not too much though (I think - I haven't tried a prototype yet), just the core.Node and related objects changing there

Reply:
I suspect the duplication would easily leak into many functions than just the handler.